diff --git a/db/Medidata.Database/StoredProcedures/spGetInstanceNameByBalanceVisitIdentifier.sql b/db/Medidata.Database/StoredProcedures/spGetInstanceNameByBalanceVisitIdentifier.sql
new file mode 100644
index 00000000000..e928558878d
--- /dev/null
+++ b/db/Medidata.Database/StoredProcedures/spGetInstanceNameByBalanceVisitIdentifier.sql
@@ -0,0 +1,24 @@
+SET ANSI_NULLS ON
+GO
+
+SET QUOTED_IDENTIFIER ON
+GO
+
+IF OBJECT_ID('dbo.spGetInstanceNameByBalanceVisitIdentifier') IS NULL
+    EXEC ('CREATE PROCEDURE dbo.spGetInstanceNameByBalanceVisitIdentifier AS RETURN')
+GO
+
+ALTER PROCEDURE [dbo].[spGetInstanceNameByBalanceVisitIdentifier] (@SubjectID INT, @BalanceVisitIdentifier UNIQUEIDENTIFIER, @Locale VARCHAR(3)) 
+AS
+BEGIN
+    SET NOCOUNT ON
+
+	SELECT dbo.fnFormatInstanceName(i.InstanceName, lds.String, i.InstanceRepeatNumber, s.SubjectDate, i.InstanceDate)
+	FROM   dbo.Instances i
+           JOIN dbo.Subjects s ON s.SubjectID = i.SubjectID
+		   JOIN dbo.Folders f ON f.FolderID = i.FolderID
+		   JOIN dbo.LocalizedDataStrings lds ON lds.StringID = f.FolderName
+	WHERE  i.SubjectID = @SubjectID
+	       AND i.BalanceVisitIdentifier = @BalanceVisitIdentifier
+		   AND lds.Locale = @Locale
+END
diff --git a/rave/Medidata.Core.Objects.Test/Services/InstanceServiceTest.cs b/rave/Medidata.Core.Objects.Test/Services/InstanceServiceTest.cs
index fa61dc367d8..ff1d567d82b 100644
--- a/rave/Medidata.Core.Objects.Test/Services/InstanceServiceTest.cs
+++ b/rave/Medidata.Core.Objects.Test/Services/InstanceServiceTest.cs
@@ -195,5 +195,62 @@ public void FetchSubjectChildInstances_CalledWithNullInteraction_ThrowsException
                 _instanceService.FetchSubjectChildInstances( Arg<int>.Is.Anything, null,
                     Arg<bool>.Is.Anything));
         }
+
+        [TestMethod]
+        public void GetInstanceNameByBalanceVisitIdentifier_NegativeSubjectId_ThrowsException()
+        {
+            //arrange
+            int subjectId = -10;
+            var balanceVisitIdentifier = Guid.NewGuid().ToString();
+            var interactionMock = _container.GetMock<IInteraction>();
+
+            //assert
+            Assert.ThrowsException<ArgumentOutOfRangeException>(
+                () => _instanceService.GetInstanceNameByBalanceVisitIdentifier(subjectId, balanceVisitIdentifier, interactionMock),
+                $"Subject ID {subjectId} is not valid.");
+        }
+
+        [TestMethod]
+        public void GetInstanceNameByBalanceVisitIdentifier_CalledWithNullInteraction_ThrowsException()
+        {
+            Assert.ThrowsException<ArgumentNullException>(() =>
+                _instanceService.GetInstanceNameByBalanceVisitIdentifier(Arg<int>.Is.Anything, Arg<Guid>.Is.Anything.ToString(),
+                    null));
+        }
+
+        [TestMethod]
+        public void GetInstanceNameByBalanceVisitIdentifier_InvalidBalanceVisitIdentifierGuid_ThrowsException()
+        {
+            //arrange
+            var subjectId = 1;
+            var balanceVisitIdentifier = "FFG10-33FEOS";
+            var interactionMock = _container.GetMock<IInteraction>();
+
+            //assert
+            Assert.ThrowsException<ArgumentException>(
+                () => _instanceService.GetInstanceNameByBalanceVisitIdentifier(subjectId, balanceVisitIdentifier, interactionMock),
+                $"Balance Visit Identifier {balanceVisitIdentifier} is not valid.");
+        }
+
+        [TestMethod]
+        public void GetInstanceNameByBalanceVisitIdentifier_IsValid()
+        {
+            //arrange
+            var subjectId = 1;
+            var balanceVisitIdentifierGuid = Guid.NewGuid();
+            var balanceVisitIdentifier = balanceVisitIdentifierGuid.ToString();
+            var interactionMock = _container.GetMock<IInteraction>();
+            interactionMock.Stub(x => x.DefaultLocale).Return("eng");
+            var expectedResult = "Folder";
+
+            _instanceRepositoryMock.Stub(x => x.GetInstanceNameByBalanceVisitIdentifier(subjectId, balanceVisitIdentifierGuid, interactionMock.DefaultLocale)).Return(expectedResult);
+
+            //act
+            var actualResult =
+                _instanceService.GetInstanceNameByBalanceVisitIdentifier(subjectId, balanceVisitIdentifier, interactionMock);
+
+            //assert
+            Assert.AreEqual(expectedResult, actualResult);
+        }
     }
 }
diff --git a/rave/Medidata.Core/Medidata.Core.Objects/Balance/eDA/RaveRepository.cs b/rave/Medidata.Core/Medidata.Core.Objects/Balance/eDA/RaveRepository.cs
index 0184c82fdad..9ec6db87648 100644
--- a/rave/Medidata.Core/Medidata.Core.Objects/Balance/eDA/RaveRepository.cs
+++ b/rave/Medidata.Core/Medidata.Core.Objects/Balance/eDA/RaveRepository.cs
@@ -22,6 +22,7 @@ public class RaveRepository : IRaveRepository
         private readonly IFormService _formService;
         private readonly IMatrixService _matrixService;
         private readonly IDataPointService _dataPointService;
+        private readonly IInstanceService _instanceService;
 
         static RaveRepository()
         {
@@ -32,13 +33,15 @@ static RaveRepository()
             ISubjectService subjectService,
             IFormService formService,
             IMatrixService matrixService,
-            IDataPointService dataPointService)
+            IDataPointService dataPointService,
+            IInstanceService instanceService)
         {
             _logger = logger;
             _subjectService = subjectService;
             _formService = formService;
             _matrixService = matrixService;
             _dataPointService = dataPointService;
+            _instanceService = instanceService;
             _logger.CurrentLoggerType = typeof(RaveRepository);
         }
 
@@ -136,9 +139,14 @@ public void SavePDALForm(IPdalForm pdalForm)
                 record.Active = true;
                 record.RequestSave();
             
-
                 logLine.LastModifiedBy = lastUpdateUser;
 
+                var subjectId = record.DataPage.Subject.ID;
+                var formName = _instanceService.GetInstanceNameByBalanceVisitIdentifier(subjectId, logLine.VisitIdentifier, pdalForm.Interaction);
+
+                if (!string.IsNullOrEmpty(formName))
+                    logLine.Folder = formName;
+
                 Mapper.Map(logLine, record, opts => opts.Items["dataPointService"] = _dataPointService);
 
                 if (logLine.Locked)
diff --git a/rave/Medidata.Core/Medidata.Core.Objects/Interfaces/Repositories/IInstanceRepository.cs b/rave/Medidata.Core/Medidata.Core.Objects/Interfaces/Repositories/IInstanceRepository.cs
index 0d3f413cc5e..7271612705e 100644
--- a/rave/Medidata.Core/Medidata.Core.Objects/Interfaces/Repositories/IInstanceRepository.cs
+++ b/rave/Medidata.Core/Medidata.Core.Objects/Interfaces/Repositories/IInstanceRepository.cs
@@ -1,4 +1,5 @@
 ﻿using Medidata.Core.Objects.DataTransferObjects;
+using System;
 using System.Collections.Generic;
 
 namespace Medidata.Core.Objects.Interfaces.Repositories
@@ -9,6 +10,6 @@ internal interface IInstanceRepository
         IEnumerable<InstanceDataDTO> GetInstancesForParentInstance(int parentInstanceId, int userId);
         IEnumerable<InstanceDataDTO> GetInstancesForSubjectIncludingDescendants(int subjectId, int userId);
         IEnumerable<InstanceDataDTO> GetInstancesForSubjectExcludingDescendants(int subjectId, int userId);
-
+        string GetInstanceNameByBalanceVisitIdentifier(int subjectId, Guid balanceVisitIdentifier, string locale);
     }
 }
diff --git a/rave/Medidata.Core/Medidata.Core.Objects/Repositories/InstanceRepository.cs b/rave/Medidata.Core/Medidata.Core.Objects/Repositories/InstanceRepository.cs
index 532b14cdd53..5ada0fe5f7f 100644
--- a/rave/Medidata.Core/Medidata.Core.Objects/Repositories/InstanceRepository.cs
+++ b/rave/Medidata.Core/Medidata.Core.Objects/Repositories/InstanceRepository.cs
@@ -2,6 +2,7 @@
 using Medidata.Core.Objects.DataTransferObjects;
 using Medidata.Core.Objects.Interfaces.Repositories;
 using Medidata.Data.Factories;
+using System;
 using System.Collections.Generic;
 
 namespace Medidata.Core.Objects.Repositories
@@ -63,5 +64,17 @@ public IEnumerable<InstanceDataDTO> GetInstancesForSubjectExcludingDescendants(
                 return instances;
             }
         }
+
+        public string GetInstanceNameByBalanceVisitIdentifier(int subjectId, Guid balanceVisitIdentifier, string locale)
+        {
+            using (var adapter = _factory.GetDataAdapter(false))
+            {
+                adapter.AddInputParameter("SubjectID", subjectId);
+                adapter.AddInputParameter("BalanceVisitIdentifier", balanceVisitIdentifier);
+                adapter.AddInputParameter("Locale", locale);
+                var instanceName = adapter.GetItem<string>("spGetInstanceNameByBalanceVisitIdentifier");
+                return instanceName;
+            }
+        }
     }
 }
diff --git a/rave/Medidata.Core/Medidata.Core.Objects/Services/InstanceService.cs b/rave/Medidata.Core/Medidata.Core.Objects/Services/InstanceService.cs
index b58bfa5c6dd..503ca17a7e5 100644
--- a/rave/Medidata.Core/Medidata.Core.Objects/Services/InstanceService.cs
+++ b/rave/Medidata.Core/Medidata.Core.Objects/Services/InstanceService.cs
@@ -74,5 +74,17 @@ public IInstances FetchSubjectChildInstances( int subjectId, IInteraction intera
             var instances = _instanceFetcher.FindOrCreateCollection(FetchData, interaction, true);
             return instances;
         }
+
+        public string GetInstanceNameByBalanceVisitIdentifier(int subjectId, string visitIdentifier, IInteraction interaction)
+        {
+            if (interaction == null)
+                throw new ArgumentNullException(nameof(interaction));
+            if (subjectId < 1)
+                throw new ArgumentOutOfRangeException(nameof(subjectId), $"Subject ID {subjectId} is not valid.");
+            if (!Guid.TryParse(visitIdentifier, out var guidVisitIdentifier))
+                throw new ArgumentException(nameof(visitIdentifier), $"Balance Visit Identifier {visitIdentifier} is not valid.");
+
+            return _repository.GetInstanceNameByBalanceVisitIdentifier(subjectId, guidVisitIdentifier, interaction.DefaultLocale);
+        }
     }
 }
diff --git a/rave/Medidata.Interfaces/Medidata.Interfaces/Adapters/IInstanceService.cs b/rave/Medidata.Interfaces/Medidata.Interfaces/Adapters/IInstanceService.cs
index 3c6c72de1ef..f3f93f43140 100644
--- a/rave/Medidata.Interfaces/Medidata.Interfaces/Adapters/IInstanceService.cs
+++ b/rave/Medidata.Interfaces/Medidata.Interfaces/Adapters/IInstanceService.cs
@@ -11,5 +11,6 @@ public interface IInstanceService
         IInstances FetchInstanceChildInstances(int parentInstanceIds, IInteraction interaction);
         IInstances FetchSubjectChildInstances(int subjectId, IInteraction interaction,
             bool includeDescendants);
+        string GetInstanceNameByBalanceVisitIdentifier(int subjectId, string visitIdentifier, IInteraction interaction);
     }
 }
diff --git a/src/Core/Objects/src/EDC/Query.Static.cs b/src/Core/Objects/src/EDC/Query.Static.cs
index 436a9a4c847..a6414c2cadc 100644
--- a/src/Core/Objects/src/EDC/Query.Static.cs
+++ b/src/Core/Objects/src/EDC/Query.Static.cs
@@ -51,7 +51,7 @@ public sealed partial class Query
         int queryTextStringID,
         bool answerOnChange,
         bool closeOnChange,
-        int checkID,
+        long checkID,
         string checkHash,
         string? answerHash = null)
         : base(
diff --git a/src/Core/Objects/src/EDC/Query.cs b/src/Core/Objects/src/EDC/Query.cs
index 81a2f6b5b63..9c7a9b2740a 100644
--- a/src/Core/Objects/src/EDC/Query.cs
+++ b/src/Core/Objects/src/EDC/Query.cs
@@ -53,7 +53,7 @@ internal Query(MarkingDependencies dependencies, MarkingDto marking, IInteractio
         int queryTextStringId,
         bool answerOnChange,
         bool closeOnChange,
-        int checkId,
+        long checkId,
         string? checkHash,
         string? answerHash = null,
         bool obeyMaxQueryLength = true)